### PR TITLE
FIX: Respect blocked domains list when redirecting

### PR DESCRIPTION
**Pull request by nattsw**
/t/59305

Our previous implementation used a simple `blocked_domain_array.include?(hostname)`
so some values were not matching. Additionally, in some configurations like ours, we'd used
"cat.*.dog.com" with the assumption we'd support globbing.

This change implicitly allows globbing by blocking "http://a.b.com" if "b.com" is a blocked 
domain but does not actively do anything for "*".

An upcoming change might include frontend validation for values that can be inserted.

Screenshot 2022-01-20 at 12 11 51 PM

### CODE REVIEW

File: lib/retrieve_title.rb [hunk 1]
```diff
@@ -60,6 +60,10 @@ def self.fetch_title(url)
     encoding = nil
 
     fd.get do |_response, chunk, uri|
+      if (uri.present? && InlineOneboxer.domain_is_blocked?(uri.hostname))
+        throw :done
```
Comment (SamSaffron) [inline lib/retrieve_title.rb:64]: tricky... but I am not sure if this is technically correct, we only want to apply the ban hammer on the last redirect in the chain, so if you work through a blocked onebox domain on your way to a non blocked one, it is fine.

File: lib/inline_oneboxer.rb [hunk 2]
```diff
@@ -74,6 +73,12 @@ def self.lookup(url, opts = nil)
 
   private
 
+  def self.domain_is_blocked?(hostname)
+    SiteSetting.blocked_onebox_domains&.split('|').any? do |blocked|
```
Comment (SamSaffron) [inline lib/inline_oneboxer.rb:77]: can this logic be reused by standard oneboxer as well?

File: lib/retrieve_title.rb (hunk 1 above)
Comment (nattsw) [reply lib/retrieve_title.rb:64]: I wrote a test and it fulfils what you said. I'll add it in as well.

The `FinalDestination` class only returns the ... final destination, and it's what we're using here.

File: lib/inline_oneboxer.rb (hunk 2 above)
Comment (nattsw) [reply lib/inline_oneboxer.rb:77]: For sure. My focus was just on titles but I should explore outwards. Will start with the usages of ... `FinalDestination`

File: lib/inline_oneboxer.rb [hunk 3]
```diff
@@ -74,6 +73,12 @@ def self.lookup(url, opts = nil)
...
+      hostname.match("(^|\\.)#{Regexp.quote(blocked)}$")
```
Comment (SamSaffron) [inline lib/inline_oneboxer.rb:78]: mixed on this ... but we could save on the regex with:

`"a.b.c".end_with?(".b.c") || "a.b.c" == ".b.c"` 

Do upper case domains get a bypass here? Can they be injected somehow?

**Review by SamSaffron — APPROVED** [review]

File: lib/inline_oneboxer.rb (hunk 3 above)
Comment (SamSaffron) [reply lib/inline_oneboxer.rb:78]: oh also if not going without regex ... be sure to use `match?` which returns a bool and is faster. 

```
> "a".match? /a/
=> true
```

Was something I pushed nobu to add :)

Comment (nattsw) [reply lib/inline_oneboxer.rb:78]: I'll have a look and see which is neater

**Pull request merged**